File: app_watcher/src/main/java/com/tongchen/twatcher/base/ui/activity/MVPActivity.java
package com.tongchen.twatcher.base.ui.activity;

import android.os.Bundle;

import com.tongchen.twatcher.base.presenter.IMVPPresenter;
import com.tongchen.twatcher.base.view.IMVPView;

import javax.inject.Inject;

/**
 * Created by TongChen at 22:43 on 2018/6/24.
 * <p>
 * Description: MVP Activity 基类
 */
public abstract class MVPActivity<E, V extends IMVPView<E>, P extends IMVPPresenter<V>>
        extends BaseActivity implements IMVPView<E> {

    @Inject
    protected P mPresenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        injectActivity();
        mPresenter.attachView((V) this);

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mPresenter.detachView();
    }

    protected abstract void injectActivity();
}

File: app_watcher/src/main/java/com/tongchen/twatcher/base/ui/activity/MVPActivity.kt
package com.tongchen.twatcher.base.ui.activity

import android.os.Bundle

import com.tongchen.twatcher.base.presenter.IMVPPresenter
import com.tongchen.twatcher.base.view.IMVPView

import javax.inject.Inject

/**
 * Created by TongChen at 22:43 on 2018/6/24.
 *
 *
 * Description: MVP Activity 基类
 */
abstract class MVPActivity<E, V : IMVPView<E>, P : IMVPPresenter<V>> : BaseActivity() {

    @Inject
    lateinit var mPresenter: P

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        injectActivity()
        mPresenter.attachView(this as V)

    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter.detachView()
    }

    protected abstract fun injectActivity()
}


